text/0093-rocksdb-per-region.md

```diff
@@ -0,0 +1,196 @@
+
+```mermaid
+flowchart TB
+    mem0-->phys0
+    mem0-->phys1
+    phys0-->|hardlink|phys2
+    phys1-->|hardlink|phys2
+    subgraph 2_5
+    mem0["Frozen memtable"]
+    phys2["sst on disk"]
+    end
+    subgraph 2_11
+    mem2["active memtable"]
+    phys0["sst on disk"]
+    end
+    subgraph 3_5
+    mem3["active memtable"]
+    phys1["sst on disk"]
+    end
+```
+
+For example, if region 2 is split into region 2 and region 3 at index 11, then create tablets 2_11 and 3_5. Before split command is persisted in all new tablets, although new writes are written to new tablets, but tablet index in raft db of region 2 should still point to the old tablet and old tablet should be kept. After all tablets have persisted the split command, region 2 can point to the new tablet, and old tablet can be removed after all queries are finished.
+
+FreezeAndClone can create multiple references to the same frozen memtables, which will cause duplicated flush. To avoid unnecessary IO, RocksDb needs to support flush filter, which will filter out all keys that are not in the current region and rebuild the bloom filter. Because split will create a new tablet, so flush filter will not corrupt existing rocksdb snapshot. After applying flush filter, every tablet will only write data in its own range, only little extra CPU cost, and no extra IO cost.
+
+FreezeAndClone will hardlink all SST files in the original tablet, so every tablet has data that is not in its own range. After split, DeleteFilesInRange can be used to quickly free hard links. And then compaction filter + manual compaction can be used to do cleanup. Also because of new tablets, existing snapshots won’t be corrupted. The cleanup can be done in background slowly.
+
+The maximum size of a tablet is supposed to be 10GiB, its file count can usually be 20 ~ 60. Hard link tens of files can only take hundreds of microseconds, which is smaller than a network roundtrip. A tablet only contains about 2 or 3 layers, cleanup may only need to rewrite about 2 or 3 files.
+
+### Merge
+
+The merge algorithm is almost the same as v6.1. To make it easy, we introduce a new interface: FreezeAndMerge. It will freeze the mutable memtables of the two tablet, and then merge their SST files by level and frozen memtables. After being merged, the new tablet need to adjust the Log Sequence Number (LSN) to the maximum number of two original tablets. To guarantee correctness, merge should be triggered only after clean up has finished in both two tablets, which should only contain the data within their region. This can be done by an additional check. For raft cf, there is no need to merge, only the target tablet’s version needs to be kept. After merge, there may be LSN regression in frozen memtables. This can only happen when two frozen memtables are from two tablets. As explained, the two memtables can’t overlap with each other, so LSN regression should not break correctness.
```

Comment:
how does it know `clean up has finished in both two tablets`?

Reply:
As said in the paragraph, "This can be done by an additional check." The simplest way is to schedule a `pre-merge` check to see if all replicas have finish cleaned up.